File: 1046.最后一块石头的重量.py
#
# @lc app=leetcode.cn id=1046 lang=python3
#
# [1046] 最后一块石头的重量
#

# @lc code=start
class Solution:
    def lastStoneWeight(self, stones):
        import heapq
        heap = []
        for i in stones:
            heapq.heappush(heap, i*-1)
        left_stones = len(stones)
        while left_stones > 1:
            first = heapq.heappop(heap)
            second = heapq.heappop(heap)
            if first != second:
                heapq.heappush(heap, first-second)
                left_stones -= 1
            else:
                left_stones -= 2
        if left_stones == 1:
            return heapq.heappop(heap) * (-1)
        else:
            return 0

# @lc code=end

File: 1046.最后一块石头的重量.java
/*
 * @lc app=leetcode.cn id=1046 lang=java
 *
 * [1046] 最后一块石头的重量
 */

// @lc code=start
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        for(int stone: stones) {
            pq.offer(stone);
        }

        while(pq.size()>=2){
            int a = pq.poll();
            int b = pq.poll();
            if(a > b) {
                pq.offer(a-b);
            }
        }

        return pq.isEmpty()?0: pq.poll();
    }
}
// @lc code=end

